File: app/src/main/java/com/example/ftmsnew/database/FTMSDatabase.java
package com.example.ftmsnew.database;

import android.content.Context;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;
import android.database.sqlite.SQLiteOpenHelper;
import android.util.Log;


import androidx.annotation.NonNull;
import androidx.annotation.Nullable;


import com.example.ftmsnew.common.CommonUtils;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.sql.SQLException;

public class FTMSDatabase extends SQLiteOpenHelper {
    public static final String LOG_TAG = FTMSDatabase.class.getName();
    public final static int DATA_VERSION = 5;
    public final static String DATABASE_NAME = "ftmsdb.sqlite";
    @NonNull
    public static String Lock = "dblock";
    public static FTMSDatabase ftmsDatabase;
    public static String DB_PATH;
    @Nullable
    public static SQLiteDatabase mSqLiteDatabase = null;
    public Context mContext;
    File rootDirectory;

    public FTMSDatabase(Context context) {
        super(context, DATABASE_NAME, null, DATA_VERSION);
        this.mContext = context;
        File dbDirectory = new File(CommonUtils.getFTMSFileRootPath() + "FTMS_Database");
        DB_PATH = dbDirectory.getAbsolutePath() + File.separator;
        Log.v("The Database Path", DB_PATH);
    }

//    public FTMSDatabase(@Nullable Context context, @Nullable String name, @Nullable SQLiteDatabase.CursorFactory factory, int version) {
//        super(context, name, factory, version);
//    }


    public static synchronized FTMSDatabase getPFtmseDatabase(Context context) {
        synchronized (Lock) {
            if (ftmsDatabase == null) {
                ftmsDatabase = new FTMSDatabase(context);
            }
            return ftmsDatabase;
        }

    }


    @Nullable
    public static SQLiteDatabase openDataBaseNew() throws SQLException {
        // Open the database
        if (mSqLiteDatabase == null) {
            mSqLiteDatabase = SQLiteDatabase.openDatabase(DB_PATH, null, SQLiteDatabase.OPEN_READWRITE | SQLiteDatabase.CREATE_IF_NECESSARY | SQLiteDatabase.NO_LOCALIZED_COLLATORS);
        } else if (!mSqLiteDatabase.isOpen()) {
            mSqLiteDatabase = SQLiteDatabase.openDatabase(DB_PATH, null, SQLiteDatabase.OPEN_READWRITE | SQLiteDatabase.CREATE_IF_NECESSARY | SQLiteDatabase.NO_LOCALIZED_COLLATORS);
        }
        return mSqLiteDatabase;
    }

    public static void copy(@NonNull File src, @NonNull File dst) throws IOException {
        InputStream in = new FileInputStream(src);
        OutputStream out = new FileOutputStream(dst);

        // Transfer bytes from in to out
        byte[] buf = new byte[1024];
        int len;
        while ((len = in.read(buf)) > 0) {
            out.write(buf, 0, len);
        }
        in.close();
        out.close();
    }

    /* create an empty database if data base is not existed */
    public void createDataBase() throws IOException {
        boolean dbExist = checkDataBase();

        if (dbExist) {
            //do nothing - database already exist
        } else {
            try {
                copyDataBase();
                Log.v("dbcopied:::", "true");
            } catch (SQLiteException ex) {
                ex.printStackTrace();
                throw new Error("Error copying database");
            } catch (IOException e) {
                e.printStackTrace();
                throw new Error("Error copying database");
            }
            try {
                openDataBase();
            } catch (SQLiteException ex) {
                ex.printStackTrace();
                throw new Error("Error opening database");
            } catch (Exception e) {
                e.printStackTrace();
                throw new Error("Error opening database");
            }
        }

    }

    /* checking the data base is existed or not */
    /* return true if existed else return false */
    public boolean checkDataBase() {
        boolean dataBaseExisted = false;
        try {
            String check_Path = DB_PATH + DATABASE_NAME;
            mSqLiteDatabase = SQLiteDatabase.openDatabase(check_Path, null, SQLiteDatabase.OPEN_READWRITE);
        } catch (Exception ex) {
            // TODO: handle exception
            ex.printStackTrace();
        }
        return mSqLiteDatabase != null;
    }

    public void copyFile(InputStream in, @NonNull OutputStream out) throws IOException {
        byte[] buffer = new byte[1024];
        int read;
        while ((read = in.read(buffer)) != -1) {
            out.write(buffer, 0, read);
        }
    }

    public void copyDataBase() throws IOException {
        File dbDir = new File(DB_PATH);
        if (!dbDir.exists()) {
            dbDir.mkdir();

        }
        InputStream myInput = mContext.getAssets().open(DATABASE_NAME);
        OutputStream myOutput = new FileOutputStream(DB_PATH + DATABASE_NAME);
        copyFile(myInput, myOutput);

    }

    public boolean UpdateGeoTagLatLng(String UpdatedByUserId,  String UpdatedDate,double Latitude,double Longitude) {


        return true;
    }


    /* Open the database */
    public void openDataBase() throws SQLException {

        String check_Path = DB_PATH + DATABASE_NAME;
        if (mSqLiteDatabase != null) {
            mSqLiteDatabase.close();
            mSqLiteDatabase = null;
            mSqLiteDatabase = SQLiteDatabase.openDatabase(check_Path, null, SQLiteDatabase.OPEN_READWRITE);
        } else {
            mSqLiteDatabase = SQLiteDatabase.openDatabase(check_Path, null, SQLiteDatabase.OPEN_READWRITE);
        }
    }

    @Override
    public void onCreate(SQLiteDatabase db) {

    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {


    }

    public void openDatabase() {
        String dbPath = mContext.getDatabasePath(DATABASE_NAME).getPath();
        if(mSqLiteDatabase != null && mSqLiteDatabase.isOpen()) {
            return;
        }
        mSqLiteDatabase = SQLiteDatabase.openDatabase(dbPath, null, SQLiteDatabase.OPEN_READWRITE);
    }




}

File: app/src/main/java/com/example/ftmsnew/database/FTMSDatabase.kt
package com.example.ftmsnew.database

import android.annotation.SuppressLint
import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteException
import android.database.sqlite.SQLiteOpenHelper
import android.util.Log
import java.math.BigInteger

class FTMSDatabase(context: Context) :
    SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {

    override fun onCreate(db: SQLiteDatabase?) {
        val createZones = "CREATE TABLE Zones (\n" +
                "   Id Integer PrimariKey Auto IncreMent,\n" +
                "    Code            VARCHAR (6)                         NOT NULL,\n" +
                "    Name            VARCHAR (100)                       NOT NULL,\n" +
                "    StateId         INT                                 NOT NULL,\n" +
                "    VillageId       INT                                 NOT NULL,\n" +
                "    IsActive        BIT                                 NOT NULL,\n" +
                "    CreatedDate     DATETIME                            NOT NULL,\n" +
                "    CreatedByUserId INT                                 NOT NULL,\n" +
                "    UpdatedDate     DATETIME                            NOT NULL,\n" +
                "    UpdatedByUserId INT                                 NOT NULL\n" +
                ");\n"
        db!!.execSQL(createZones)

        val createVillage = "CREATE TABLE Village (\n" +
                "   Id Integer PrimariKey Auto IncreMent,\n" +
                "    Code            VARCHAR (6)                         NOT NULL,\n" +
                "    Name            VARCHAR (100)                       NOT NULL,\n" +
                "    MandalId        INT                                 NOT NULL,\n" +
                "    IsActive        BIT                                 NOT NULL,\n" +
                "    CreatedDate     DATETIME                            NOT NULL,\n" +
                "    CreatedByUserId INT                                 NOT NULL,\n" +
                "    UpdatedDate     DATETIME                            NOT NULL,\n" +
                "    UpdatedByUserId INT                                 NOT NULL\n" +

                ");\n"
        db.execSQL(createVillage)

        val createStaes = "CREATE TABLE States(\n" +
                "Id Integer PrimariKey Auto IncreMent\n" +
                "Code varchar(6) NOT NULL,\n" +
                "Name varchar(100) NOT NULL,\n" +
                "IsActive bit NOT NULL,\n" +
                "CreatedDate datetime NOT NULL,\n" +
                "CreatedByUserId int NOT NULL,\n" +
                "UpdatedDate datetime NOT NULL,\n" +
                "UpdatedByUserId int NOT NULL)\n"
        db.execSQL(createStaes)

        val createResorchStation = "\n" +
                "CREATE TABLE ResearchStations(\n" +
                "Id Integer PrimariKey Auto IncreMent,\n" +
                "Code varchar(6) NOT NULL,\n" +
                "Name varchar(100) NOT NULL,\n" +
                "ZoneId int NOT NULL,\n" +
                "VillageId int NOT NULL,\n" +
                "MobileNumber bigint NOT NULL,\n" +
                "Address varchar(250) NOT NULL,\n" +
                "IsActive bit NOT NULL,\n" +
                "CreatedDate datetime NOT NULL,\n" +
                "CreatedByUserId int NOT NULL,\n" +
                "UpdatedDate datetime NOT NULL,\n" +
                "UpdatedByUserId int NOT NULL)"
        db.execSQL(createResorchStation)
        val createMandal = "CREATE TABLE Mandal(\n" +
                "Id Integer PrimariKey Auto IncreMent,\n" +
                "Code varchar(6) NOT NULL,\n" +
                "Name varchar(100) NOT NULL,\n" +
                "DistrictId int NOT NULL,\n" +
                "IsActive bit NOT NULL,\n" +
                "CreatedDate datetime NOT NULL,\n" +
                "CreatedByUserId int NOT NULL,\n" +
                "UpdatedDate datetime NOT NULL,\n" +
                "UpdatedByUserId int NOT NULL)\n"
        db.execSQL(createMandal)
        val createDistrict = "\n" +
                "CREATE TABLE District(\n" +
                "Id Integer PrimariKey Auto IncreMent,\n" +
                "Code varchar(6) NOT NULL,\n" +
                "Name varchar(100) NOT NULL,\n" +
                "ZoneId int NOT NULL,\n" +
                "IsActive bit NOT NULL,\n" +
                "CreatedDate datetime NOT NULL,\n" +
                "CreatedByUserId int NOT NULL,\n" +
                "UpdatedDate datetime NOT NULL,\n" +
                "UpdatedByUserId int NOT NULL)"
        db.execSQL(createDistrict)
        val createCrop = "CREATE TABLE Crop(\n" +
                "Id Integer PrimariKey Auto IncreMent,\n" +
                "Code varchar(6) NOT NULL,\n" +
                "Name varchar(50) NOT NULL,\n" +
                "IsActive bit NOT NULL,\n" +
                "CreatedByUserId int NOT NULL,\n" +
                "CreatedDate datetime NOT NULL,\n" +
                "UpdatedByUserId int NOT NULL,\n" +
                "UpdatedDate datetime NOT NULL)\n"
        db.execSQL(createCrop)
    }

    override fun onUpgrade(p0: SQLiteDatabase?, p1: Int, p2: Int) {
    }


    companion object {
        private val DATABASE_VERSION = 1
        private val DATABASE_NAME = "FTMSDb"

    }
}